### ISSUE TITLE
Testing should not begin until element fixtures are ready.

### ISSUE THREAD

**Issue by indolering** [Polymer/web-component-tester#188]
I have an element that goes through some startup steps but the test environment initiates the test before the element has finished setting up.  I ran into this error when the setup was done in a created callback.  Currently, I'm attempting to performing the setup in an anonymous function that sets a parameter's default value:

```
client: {
        type: Object,
        value: function(){
          //do suff
          return object;
        }
      }
```

Ideally, testing should not begin until the elements have been fully constructed.  I'm not sure if there is a nice event we can watch for, but perhaps we could wait until 100ms after the ready function has been called?

**Comment by indolering** [comment]
Maybe I should suggest Polymer itself should offer some sort of "done" callback for the various lifecycle events?

**Comment by jmalonzo** [comment]
@indolering We've had the same issue but you can workaround this by putting your polymer/webcomponent custom element inside a template. So, you set up your environment (e.g. in beforeEach) then add the template to the DOM once you're done. Then run your tests as usual. HTH.

**Comment by indolering** [comment]
Not sure how to add the template to the DOM when I'm done, are you suggesting using JS to create the elements imperatively?

**Comment by jmalonzo** [comment]
@indolering In your -test.html you can have something like:

```
<template>
    <my-element id="fixture"></my-element>
</template>
```
Then import the template (e.g. via document.importNode) in your beforeEach() once you're done setting up your environment. HTH.

**Comment by indolering** [comment]
@jmalonzo WRT resolving this ticket, I think your workaround either needs to be documented or you should add the capability to prevent tests from running until elements have finished instantiating.

**Comment by jmalonzo** [comment]
@indolering I think this ticket can serve as a documentation :) You can raise a PR and it to the README if you like. The workaround is pretty straightforward and I can imagine preventing tests from running until all elements have been attached requires more work.

**Comment by indolering** [comment]
@jmalonzo, I tried to follow your advice but I couldn't get it to work.  I honestly don't quite understand what's going on with importNode, could you post an example of how you would import the fixtures?

**Comment by indolering** [comment]
And, TBH, "preventing tests from running until all elements have been attached requires more work" is sort-of a copout.  I mean, you are asking everyone to perform a special setup instead of just doing it once for them : )

**Comment by jmalonzo** [comment]
@indolering I never asked anyone to do anything, and I have no obligation to you. I'm just trying to help you out, giving you advice based on my experience.

**Comment by indolering** [comment]
@jmalonzo I didn't mean to come across as a demand or anything, it just feels like default behavior violates expectations.....

Would you mind giving me a more detailed example?  I promise I'll submit a PR for the readme.md : )

**Comment by justinfagnani** [comment]
@indolering quick comment for now: can you try using `<test-fixture>` and see if that helps with timing of tests?

**Comment by indolering** [comment]
@justinfagnani I can't even remember what test this was used in.  If test-fixture theoretically fixes it, why don't we close the issue and reopen if someone comes across it in the future?

**Comment by justinfagnani** [comment]
Sounds good for now. Thanks!

**Comment by indolering** [comment]
Ha, I just tried this again and it turns it out that the test is running before a `dom-if` statement completed.  WCT needs to, by default, wait for all of the child elements to finish stamping themselves out.  Even then an element may have setup steps that must be completed before the test is run.  You need to somehow allow us to create event listeners before `<test-fixture>` has reset the element.

Polymer also needs to create a standard interface for allowing us to signal when the element has completed setting itself up.

**Comment by indolering** [comment]
@justinfagnani This should probably be split into a new ticket for some sort of setup function.

It would also be cool if `<test-fixture>` scanned the for `<dom-if>` and `<dom-repeat>` elements and fired only after they had finished stamping things out....

**Comment by bennypowers** [comment]
Coming late to the thread, I had a similar problem that was caused by putting this in my fixture:

```html
    <test-fixture id="configured">
      <template>
        <!-- NB: arrays passed in as attributes in html must be valid JSON -->
        <x-eve-greeting greetings="['h', 'i']"></x-eve-greeting>
      </template>
    </test-fixture>
```

the solution is to use single quotes around the attributes and double quotes in JSON

```html
    <test-fixture id="configured">
      <template>
        <!-- NB: arrays passed in as attributes in html must be valid JSON -->
        <x-eve-greeting greetings='["h", "i"]'></x-eve-greeting>
      </template>
    </test-fixture>
```
